Microsoft.Toolkit.HighPerformance/Helpers/Internals/SpanHelper.Hash.cs

```diff
@@ -0,0 +1,324 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.Contracts;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Toolkit.HighPerformance.Helpers.Internals
+{
+    /// <summary>
+    /// Helpers to process sequences of values by reference.
+    /// </summary>
+    internal static partial class SpanHelper
+    {
+        /// <summary>
+        /// Calculates the djb2 hash for the target sequence of items of a given type.
+        /// </summary>
+        /// <typeparam name="T">The type of items to hash.</typeparam>
+        /// <param name="r0">The reference to the target memory area to hash.</param>
+        /// <param name="length">The number of items to hash.</param>
+        /// <returns>The Djb2 value for the input sequence of items.</returns>
+        [Pure]
+#if NETCOREAPP3_1
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+#endif
+        public static unsafe int GetDjb2HashCode<T>(ref T r0, IntPtr length)
+            where T : notnull
+        {
+            int hash = 5381;
+
+            IntPtr offset = default;
+
+            while ((byte*)length >= (byte*)8)
```

Comment:
Not going to vectorize this?

Reply:
I actually wasn't sure on how to do this. I do have a vectorized logic below, which is slightly similar with the caveat that it isn't guaranteed to be equal across different devices, as the size of the SIMD registers will influence the result:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/f4eb85414ac83a84a10b22f8bcb45fd882a95d1c/Microsoft.Toolkit.HighPerformance/Helpers/Internals/SpanHelper.Hash.cs#L95

The reason why I didn't know how to vectorize this one was that each partial hash value depends on the previous value, so I thought there was no other way other than just doing the whole thing sequentially, one element at a time 🤔
I might very well be missing something here, how would you go about vectorizing this?

Reply:
Ah right, my eyes completely glazed over that `hash` was an input and the output.

Reply:
Ahahahah no problem, glad to hear the code itself is ok then 😄